src/org/traccar/reports/ReportUtils.java

```diff
@@ -250,100 +259,65 @@ private static boolean isMoving(ArrayList<Position> positions, int index,
         }
     }
 
-    public static Collection<BaseReport> detectTripsAndStops(TripsConfig tripsConfig, boolean ignoreOdometer,
-            double speedThreshold, Collection<Position> positionCollection, boolean trips) {
-
-        Collection<BaseReport> result = new ArrayList<>();
+    public static <T extends BaseReport> Collection<T> detectTripsAndStops(Collection<Position> positionCollection,
+            TripsConfig tripsConfig, boolean ignoreOdometer, double speedThreshold, Class<T> reportClass) {
+        Collection<T> result = new ArrayList<>();
 
         ArrayList<Position> positions = new ArrayList<>(positionCollection);
         if (positions != null && !positions.isEmpty()) {
-            int previousStartParkingIndex = 0;
-            int startParkingIndex = -1;
-            int previousEndParkingIndex = 0;
-            int endParkingIndex = 0;
-
-            boolean isMoving = false;
-            boolean isLast = false;
-            boolean skipped = false;
-            boolean tripFiltered = false;
-
+            boolean trips = reportClass.equals(TripReport.class);
+            MotionEventHandler  motionHandler = new MotionEventHandler(tripsConfig);
+            DeviceState deviceState = new DeviceState();
+            deviceState.setMotionState(isMoving(positions, 0, tripsConfig, speedThreshold));
+            int startTripIndex = trips && deviceState.getMotionState() ? 0 : -1;
```

Comment:
You have combined two methods without actually combining logic. If you pick some more generic names, you won't need to check "trips" variable everywhere and ideally you will only need half lines of code.

You can use something like `startEventIndex` and `startNoEventIndex`. In case of trips "event" means trip and "no event" means parking, and vice versa in case of parking.